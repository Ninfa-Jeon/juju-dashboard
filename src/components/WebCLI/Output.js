import React, { useMemo } from "react";

// Colors taken from the VSCode section of
// https://en.wikipedia.org/wiki/ANSI_escape_code#3/4_bit
const ansiColors = {
  0: null, // reset
  30: "0,0,0", // Black
  31: "205,49,49", // Red
  32: "13,188,121", // Green
  33: "229,229,16", // Yellow
  34: "26,114,200", //  Blue
  35: "188,63,188", // Magenta
  36: "17,168,205", // Cyan
  37: "229,229,229", // White
  90: "102,102,102", // Bright Black (Gray)
  91: "241,76,76", // Bright Red
  92: "35,209,139", // Bright Green
  93: "245,245,67", // Bright Yellow
  94: "59,142,234", // Bright Blue
  95: "214,112,214", // Bright Magenta
  96: "41,184,219", // Bright Cyan
  97: "229,229,229", // Bright White
};

const findANSICode = /\[\d{1,3}m/g;

const getStyle = (ansiCode) => {
  const fgColor = ansiColors[ansiCode];
  if (fgColor) {
    return `color: rgb(${fgColor})`;
  }
  // We may have been provided a background color.
  const bgColor = ansiColors[ansiCode - 10];
  if (bgColor) {
    return `color: rgb(${bgColor})`;
  }
};

const colorize = (content) => {
  const colors = Array.from(content.matchAll(findANSICode));
  if (colors.length === 0) {
    return content;
  }

  let colorizedContent = "";
  let previousIndex = 0;
  colors.forEach((color, index) => {
    const ansiCode = color[0];
    const ansiCodeNumber = ansiCode.replace("[", "").replace("m", "");

    if (color.index !== 0 && previousIndex === 0) {
      // Add the content up until the first colour without wrapping it.
      colorizedContent =
        colorizedContent + content.substring(previousIndex, color.index);
      previousIndex = color.index;
    }
    const endIndex = colors[index + 1]?.index || content.length;
    let part = content.substring(color.index, endIndex).replace(ansiCode, "");
    const style = getStyle(ansiCodeNumber);
    if (style) {
      part = `<span style="${style}">${part}</span>`;
    }
    colorizedContent = colorizedContent + part;
    previousIndex = color.index;
  });
  return colorizedContent;
};

const WebCLIOutput = ({ content }) => {
  // Strip any color escape codes from the content.
  content = content.replaceAll("\u001b", "");
  const colorizedContent = useMemo(() => colorize(content), [content]);
  if (!colorizedContent) {
    return null;
  }
  return (
    <div className="webcli__output">
      <pre>
        <code dangerouslySetInnerHTML={{ __html: colorizedContent }}></code>
      </pre>
    </div>
  );
};

export default WebCLIOutput;
