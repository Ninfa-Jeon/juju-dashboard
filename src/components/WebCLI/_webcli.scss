@import "../../scss/settings";
@import "vanilla-framework/scss/vanilla";

.webcli {
  bottom: 0;
  position: fixed;
  width: calc(100vw - #{$sidebar-collapsed});

  &__input {
    background-color: black;
    display: flex;
    height: 40px;
    padding: 0 1rem;

    &-prompt {
      color: #39b54a;
      cursor: default;
      flex-grow: 0;
      // https://github.com/canonical-web-and-design/vanilla-framework/pull/3370
      // XXX The unquote will need to be removed with the next vanilla release
      font-family: unquote($font-monospace);
      font-weight: bold;
      line-height: 40px;
      padding-right: 0.5rem;
    }

    form {
      flex-grow: 1;
    }

    &-input[type="text"] {
      background-color: black;
      color: $color-mid-light;
      // https://github.com/canonical-web-and-design/vanilla-framework/pull/3370
      // XXX The unquote will need to be removed with the next vanilla release
      font-family: unquote($font-monospace);
      height: 40px;
      margin-bottom: 0;
      padding: 0 0.5rem;
    }

    &-help {
      flex-grow: 0;
      line-height: 40px;
      margin-left: 1rem;

      i {
        cursor: pointer;
      }
    }
  }

  &__output {
    background-color: #2c001e;
    height: 300px;

    code,
    pre {
      background-color: #2c001e;
      border: none;
      color: white;
    }
  }
}
