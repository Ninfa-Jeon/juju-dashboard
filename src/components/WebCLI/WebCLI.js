import React, { useCallback, useEffect, useRef, useState } from "react";

import WebCLIOutput from "./Output.js";

import "./_webcli.scss";

const DEFAULT_PLACEHOLDER = "enter command";

const generateAddress = (controllerWSHost, modelUUID) => {
  if (!controllerWSHost || !modelUUID) {
    return null;
  }
  return `wss://${controllerWSHost}/model/${modelUUID}/commands`;
};

const WebCLI = ({ controllerWSHost, credentials, modelUUID }) => {
  const [connection, setConnection] = useState(null);
  const [placeholder, setPlaceholder] = useState(DEFAULT_PLACEHOLDER);
  const inputRef = useRef();
  const wsMessageStore = useRef();
  let [output, setOutput] = useState("");

  const setDisconnectedPlaceholder = () => {
    setPlaceholder("no web cli backend available");
  };

  const setConnectedPlaceholder = () => {
    setPlaceholder(DEFAULT_PLACEHOLDER);
  };

  /*
    The messageBuffer is required because the websocket returns messages much
    faster than React wants to update the component. Doing this allows us to
    store the messages in a buffer and then set the output every cycle.
  */
  const messageBuffer = useCallback(
    (message) => {
      wsMessageStore.current = wsMessageStore.current + message;
      setTimeout(() => {
        setOutput(wsMessageStore.current);
      });
    },
    [setOutput]
  );

  const clearMessageBuffer = () => {
    wsMessageStore.current = "";
    setOutput(""); // Clear the output when sending a new message.
  };

  const handleCommandSubmit = (e) => {
    e.preventDefault();
    clearMessageBuffer();
    connection.send(
      JSON.stringify({
        user: credentials.user,
        credentials: credentials.password,
        commands: [e.currentTarget.children.command.value],
      })
    );
    inputRef.current.value = ""; // Clear the input after sending the message.
  };

  const showHelp = () => {
    setOutput(
      `Welcome to the Juju Web CLI - see the <a href="https://juju.is/docs/webcli" target="_blank">full documentation here</a>.`
    );
  };

  const handleWSMessage = useCallback(
    (e) => {
      try {
        const data = JSON.parse(e.data);
        if (data.done) {
          // This is the last message.
          return;
        }
        if (!data.output) {
          // This is the first message, an empty object and a newline.
          return;
        }
        messageBuffer(`\n${data?.output[0]}`);
      } catch (e) {
        console.log(e);
        // XXX handle the invalid data response
      }
    },
    [messageBuffer]
  );

  useEffect(() => {
    const address = generateAddress(controllerWSHost, modelUUID);
    if (!address) {
      setDisconnectedPlaceholder();
      return;
    } else {
      setConnectedPlaceholder();
    }
    const ws = new WebSocket(address);
    ws.onopen = () => {
      setConnection(ws);
    };
    ws.onclose = setDisconnectedPlaceholder;
    ws.onmessage = handleWSMessage;
    return () => {
      ws.close();
    };
  }, [controllerWSHost, modelUUID, handleWSMessage]);

  return (
    <div className="webcli">
      <WebCLIOutput content={output} />
      <div className="webcli__input">
        <div className="webcli__input-prompt">$ juju</div>
        <form onSubmit={handleCommandSubmit}>
          <input
            className="webcli__input-input"
            type="text"
            name="command"
            ref={inputRef}
            placeholder={placeholder}
          />
        </form>
        <div className="webcli__input-help">
          <i
            className="p-icon--help is-light"
            onClick={showHelp}
            onKeyDown={showHelp}
            role="button"
            tabIndex="0"
          />
        </div>
      </div>
    </div>
  );
};

export default WebCLI;
