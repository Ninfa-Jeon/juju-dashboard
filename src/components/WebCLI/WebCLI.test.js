import React from "react";
import { mount } from "enzyme";
import { act } from "react-dom/test-utils";
import WS from "jest-websocket-mock";

import WebCLI from "./WebCLI";

describe("WebCLI", () => {
  it("renders correctly", () => {
    const wrapper = mount(<WebCLI />);
    expect(wrapper).toMatchSnapshot();
  });

  it("shows the help in the output when the ? is clicked", () => {
    const wrapper = mount(<WebCLI />);
    act(() => {
      wrapper.find(".webcli__input-help i").simulate("click");
    });
    return new Promise((resolve) => setTimeout(resolve)).then(() => {
      act(() => {
        wrapper.update();
      });
      expect(
        wrapper
          .find(".webcli__output-content code")
          .prop("dangerouslySetInnerHTML")["__html"]
      ).toBe(
        `Welcome to the Juju Web CLI - see the <a href="https://juju.is/docs/webcli" class="p-link--inverted" target="_blank">full documentation here</a>.`
      );
    });
  });

  it("shows a disconnected message if no ws is connected", () => {
    const wrapper = mount(<WebCLI />);
    expect(wrapper.find(".webcli__input-input").prop("placeholder")).toBe(
      "no web cli backend available"
    );
  });

  describe("WebCLI Output", () => {
    it("displays messages recieved over the websocket", () => {
      // ..until it receives a 'done' message.
      const server = new WS("ws://localhost:1234/model/abc123/commands", {
        jsonProtocol: true,
      });
      const wrapper = mount(
        <WebCLI
          protocol="ws"
          controllerWSHost="localhost:1234"
          credentials={{
            user: "spaceman",
            password: "somelongpassword",
          }}
          modelUUID="abc123"
        />
      );
      return new Promise(async (resolve) => {
        await act(async () => {
          await server.connected;
          wrapper.find(".webcli__input-input").instance().value =
            "status --color";
          wrapper.find("form").simulate("submit", { preventDefault: () => {} });
          await expect(server).toReceiveMessage({
            user: "spaceman",
            credentials: "somelongpassword",
            commands: ["status --color"],
          });

          const messages = [
            {
              output: [
                "Model       Controller       Cloud/Region     Version    SLA          Timestamp",
              ],
            },
            {
              output: [
                "controller  google-us-east1  google/us-east1  2.9-beta1  unsupported  17:44:14Z",
              ],
            },
            { output: [""] },
            {
              output: [
                "Machine  State    DNS             Inst id        Series  AZ          Message",
              ],
            },
            {
              output: [
                "0        started  35.190.153.209  juju-3686b9-0  focal   us-east1-b  RUNNING",
              ],
            },
            { output: [""] },
            { done: true },
          ];
          messages.forEach((message) => server.send(message));

          setTimeout(() => {
            act(() => {
              wrapper.update();
            });
            expect(
              wrapper
                .find(".webcli__output-content code")
                .prop("dangerouslySetInnerHTML")
            ).toMatchSnapshot();
            resolve();
          });
        });
      });
    });

    // it("can be resized by dragging");
    // it("displays ansi colored content colored");
    // it("displays content in the default height");
  });
});
