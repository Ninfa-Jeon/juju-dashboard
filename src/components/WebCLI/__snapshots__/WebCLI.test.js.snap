// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebCLI WebCLI Output displays ansi colored content colored 1`] = `
Object {
  "__html": "
Model       Controller       Cloud/Region     Version    SLA          Timestamp
controller  google-us-east1  google/us-east1  2.9-beta1  unsupported  17:44:14Z

Machine  State    DNS             Inst id        Series  AZ          Message
0        <span style=\\"color: rgb(13,188,121)\\">started  </span>35.190.153.209  juju-3686b9-0  focal   us-east1-b  RUNNING
",
}
`;

exports[`WebCLI WebCLI Output displays messages recieved over the websocket 1`] = `
Object {
  "__html": "
Model       Controller       Cloud/Region     Version    SLA          Timestamp
controller  google-us-east1  google/us-east1  2.9-beta1  unsupported  17:44:14Z

Machine  State    DNS             Inst id        Series  AZ          Message
0        started  35.190.153.209  juju-3686b9-0  focal   us-east1-b  RUNNING
",
}
`;

exports[`WebCLI renders correctly 1`] = `
<WebCLI
  controllerWSHost="jimm.jujucharms.com:443"
  modelUUID="abc123"
>
  <div
    className="webcli"
  >
    <WebCLIOutput
      content=""
      helpMessage="Welcome to the Juju Web CLI - see the <a href=\\"https://juju.is/docs/webcli\\" class=\\"p-link--inverted\\" target=\\"_blank\\">full documentation here</a>."
      setShouldShowHelp={[Function]}
      showHelp={false}
    >
      <div
        className="webcli__output"
        style={
          Object {
            "height": "1px",
          }
        }
      >
        <div
          aria-hidden={true}
          className="webcli__output-dragarea"
        >
          <div
            className="webcli__output-handle"
          />
        </div>
        <pre
          className="webcli__output-content"
          style={
            Object {
              "height": "1px",
            }
          }
        >
          <code
            dangerouslySetInnerHTML={
              Object {
                "__html": "",
              }
            }
          />
        </pre>
      </div>
    </WebCLIOutput>
    <div
      className="webcli__input"
    >
      <div
        className="webcli__input-prompt"
      >
        $ juju
      </div>
      <form
        onSubmit={[Function]}
      >
        <input
          autoCapitalize="none"
          autoComplete="off"
          autoCorrect="off"
          className="webcli__input-input"
          name="command"
          placeholder="enter command"
          spellCheck="false"
          type="text"
        />
      </form>
      <div
        className="webcli__input-help"
      >
        <i
          className="p-icon--help is-light"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          tabIndex="0"
        />
      </div>
    </div>
  </div>
</WebCLI>
`;
